Reserve 'c' for the client's receiver name

ikawaha opened the pull request:
Fix #2138.

# Design
```
var _ = Service("calc", func() {
	Description("The calc service performs operations on numbers.")

	Method("add", func() {
		Payload(func() {
			Field(1, "a", Int, "Left operand")
			Field(2, "c", Int, "Right operand")
		})

		Result(Int)

		HTTP(func() {
			GET("/add/{a}/{c}")  // ← ★
		})
	})
})

```
# Generated code
```go
// BuildAddRequest instantiates a HTTP request object with method and path set
// to call the "calc" service "add" endpoint
func (c *Client) BuildAddRequest(ctx context.Context, v interface{}) (*http.Request, error) {
	var (
		a  int
		c2 int // ← ★
	)
	{
		p, ok := v.(*calc.AddPayload)
		if !ok {
			return nil, goahttp.ErrInvalidType("calc", "add", "*calc.AddPayload", v)
		}
		if p.A != nil {
			a = *p.A
		}
		if p.C != nil {
			c2 = *p.C
		}
	}
	u := &url.URL{Scheme: c.scheme, Host: c.host, Path: AddCalcPath(a, c2)}
	req, err := http.NewRequest("GET", u.String(), nil)
	if err != nil {
		return nil, goahttp.ErrInvalidURL("calc", "add", u.String(), err)
	}
	if ctx != nil {
		req = req.WithContext(ctx)
	}

	return req, nil
}

```

ikawaha committed:
Reserve 'c' for the client's receiver name

ikawaha committed:
go fmt

ikawaha committed:
Add optional context parameter to RecursiveValidationCode function

ikawaha committed:
Use original path param name

ikawaha committed:
Update

ikawaha commented on http/codegen/service_data.go:
```diff
@@ -2184,7 +2191,7 @@ func extractPathParams(a *expr.MappedAttributeExpr, service *expr.AttributeExpr,
 			Map:            false,
 			MapStringSlice: false,
-			Validate:       codegen.RecursiveValidationCode(c, ctx, true, varn),
+			Validate:       codegen.RecursiveValidationCode(c, ctx, true, varn, name),
```
Setting original path param name, to avoid setting escaped name to the validation target.

before:
```
err = goa.MergeErrors(err, goa.ValidatePattern("c2", c2, "patternc")) 
```
after:
```
err = goa.MergeErrors(err, goa.ValidatePattern("c", c2, "patternc"))
```

---

There are several places where it is better to use the original name.
If this PR is adopted, correct them as well.
```
$ git grep codegen.RecursiveValidationCode|grep varn
grpc/codegen/service_data.go:			Validate:     codegen.RecursiveValidationCode(c, ctx, required, varn),
http/codegen/service_data.go:					Validate:       codegen.RecursiveValidationCode(pAtt, httpsvrctx, required, varn),
http/codegen/service_data.go:			Validate:     codegen.RecursiveValidationCode(c, ctx, required, varn),
http/codegen/service_data.go:			Validate:      codegen.RecursiveValidationCode(hattr, svcCtx, required, varn),
```

nitinmohan87 reviewed (changes requested):
Thanks for the PR!

Since you introduced a new parameter to `RecursiveValidationCode` this becomes an interface-breaking change. I am thinking if we can try to avoid making interface-breaking change for this particular bug.

Only the `BuildXXXRequest` gets this problem because of the client handle variable `c`. So we can simply modify [this particular section](https://github.com/goadesign/goa/blob/v3/http/codegen/service_data.go#L763-L805) to not conflict with the client handle variable, instead of doing it all over the generated code. I think it is just about [checking if any of the `routes[0].PathInit.ClientArgs` has a name `c`](https://github.com/goadesign/goa/blob/v3/http/codegen/service_data.go#L772-L776) and producing a new name for it. Also, we may need to modify [this template](https://github.com/goadesign/goa/blob/v3/http/codegen/service_data.go#L2562-L2605) to use `.Args` instead of `.PathInit.ClientArgs` where necessary. I think this will eliminate any interface-breaking changes. Please let me know if that makes sense!

ikawaha committed:
Rollback

Revert following commits:
05bf59f0e33e553a07e0288b51fb3e04c1663bd9 Update
d35a12ee89a11e66d7984ae9ab692bb8bff7582e Use original path param name
7a0c5cedf8989f30629411a652a59ae1e50c3c97 Add optional context parameter to RecursiveValidationCode function
a8b9a221e5f61ff2f95520f2f99d15083ada6481 go fmt
f498a1607d1f65e833ab0bcac1893795f068a113 Reserve 'c' for the client's receiver name

This reverts commit f498a1607d1f65e833ab0bcac1893795f068a113.

ikawaha committed:
Fix variable names in BuildXXXRequest

ikawaha committed:
Use .Args instead of .PathInit.ClientArgs

ikawaha commented:
Thank you for your review!
I tried to fix it in the way you suggested. I hope it makes sense.
There may be a better way. If so please close this pr.
Thank you.

---

Example desgin
```go

package design

import (
   . "goa.design/goa/v3/dsl"
)

var _ = API("calc", func() {
   Title("Calculator Service")
   Description("Service for adding numbers, a Goa teaser")
   Server("calc", func() {
      Host("localhost", func() {
         URI("http://localhost:8000")
      })
   })
})

var _ = Service("calc", func() {
   Description("The calc service performs operations on numbers.")

   Method("add", func() {
      Payload(func() {
         Field(1, "a", Int)
         Field(5, "b", Int)
         Field(2, "c", Int)
         Field(3, "c1", Int)
         Field(4, "c_", Int)
      })
      Result(Int)
      HTTP(func() {
         GET("/add/{a}/{b}/{c}/{c1}/{c_}")  // ← ★
      })
   })
})
```

Generated code
```go

// BuildAddRequest instantiates a HTTP request object with method and path set
// to call the "calc" service "add" endpoint
func (c *Client) BuildAddRequest(ctx context.Context, v interface{}) (*http.Request, error) {
   var (
      a   int
      b   int
      c_  int
      c1  int
      c_2 int
   )
   {
      p, ok := v.(*calc.AddPayload)
      if !ok {
         return nil, goahttp.ErrInvalidType("calc", "add", "*calc.AddPayload", v)
      }
      if p.A != nil {
         a = *p.A
      }
      if p.B != nil {
         b = *p.B
      }
      if p.C != nil {
         c_ = *p.C
      }
      if p.C1 != nil {
         c1 = *p.C1
      }
      if p.C != nil {
         c_2 = *p.C
      }
   }
   u := &url.URL{Scheme: c.scheme, Host: c.host, Path: AddCalcPath(a, b, c_, c1, c_2)}
   req, err := http.NewRequest("GET", u.String(), nil)
   if err != nil {
      return nil, goahttp.ErrInvalidURL("calc", "add", u.String(), err)
   }
   if ctx != nil {
      req = req.WithContext(ctx)
   }

   return req, nil
}
```

nitinmohan87 commented on http/codegen/service_data.go:
```diff
@@ -769,8 +769,12 @@ func (d ServicesData) analyze(hs *expr.HTTPServiceExpr) *ServiceData {
+				s.Unique("c") // 'c' is reserved as the client's receiver name.
 				for _, ca := range routes[0].PathInit.ClientArgs {
 					if ca.FieldName != "" {
+						ca.Name = s.Unique(ca.Name, "_")
```
Instead of using `Unique` you can use `Name` which will automatically suffix a counter variable to the name. You don't have to suffix with `_`
```
ca.Name = s.Name(ca.Name)
```

nitinmohan87 commented:
Looks great! One minor comment and then this is good to go.

ikawaha committed:
Remove the suffix argument

raphael commented:
Thank you!

raphael commented:
@ikawaha do you think you could backport this fix to `v2`? (simply cherry-pick the merge commit into a branch created off of `v2`). Thank you!
